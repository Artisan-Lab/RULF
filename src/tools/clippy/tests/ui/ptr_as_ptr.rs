// run-rustfix
// aux-build:macro_rules.rs

#![warn(clippy::ptr_as_ptr)]
#![feature(custom_inner_attributes)]

extern crate macro_rules;

macro_rules! cast_it {
    ($ptr: ident) => {
        $ptr as *const i32
    };
}

fn main() {
    let ptr: *const u32 = &42_u32;
    let mut_ptr: *mut u32 = &mut 42_u32;

    let _ = ptr as *const i32;
    let _ = mut_ptr as *mut i32;

    // Make sure the lint can handle the difference in their operator precedences.
    unsafe {
        let ptr_ptr: *const *const u32 = &ptr;
        let _ = *ptr_ptr as *const i32;
    }

    // Changes in mutability. Do not lint this.
    let _ = ptr as *mut i32;
    let _ = mut_ptr as *const i32;

    // `pointer::cast` cannot perform unsized coercions unlike `as`. Do not lint this.
    let ptr_of_array: *const [u32; 4] = &[1, 2, 3, 4];
    let _ = ptr_of_array as *const [u32];
    let _ = ptr_of_array as *const dyn std::fmt::Debug;

    // Ensure the lint doesn't produce unnecessary turbofish for inferred types.
    let _: *const i32 = ptr as *const _;
    let _: *mut i32 = mut_ptr as _;

    // Make sure the lint is triggered inside a macro
    let _ = cast_it!(ptr);

    // Do not lint inside macros from external crates
    let _ = macro_rules::ptr_as_ptr_cast!(ptr);
}

fn _msrv_1_37() {
    #![clippy::msrv = "1.37"]
    let ptr: *const u32 = &42_u32;
    let mut_ptr: *mut u32 = &mut 42_u32;

    // `pointer::cast` was stabilized in 1.38. Do not lint this
    let _ = ptr as *const i32;
    let _ = mut_ptr as *mut i32;
}

fn _msrv_1_38() {
    #![clippy::msrv = "1.38"]
    let ptr: *const u32 = &42_u32;
    let mut_ptr: *mut u32 = &mut 42_u32;

    let _ = ptr as *const i32;
    let _ = mut_ptr as *mut i32;
}
