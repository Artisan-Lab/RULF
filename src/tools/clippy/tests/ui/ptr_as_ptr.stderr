error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:19:13
   |
LL |     let _ = ptr as *const i32;
   |             ^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `ptr.cast::<i32>()`
   |
   = note: `-D clippy::ptr-as-ptr` implied by `-D warnings`

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:20:13
   |
LL |     let _ = mut_ptr as *mut i32;
   |             ^^^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `mut_ptr.cast::<i32>()`

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:25:17
   |
LL |         let _ = *ptr_ptr as *const i32;
   |                 ^^^^^^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `(*ptr_ptr).cast::<i32>()`

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:38:25
   |
LL |     let _: *const i32 = ptr as *const _;
   |                         ^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `ptr.cast()`

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:39:23
   |
LL |     let _: *mut i32 = mut_ptr as _;
   |                       ^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `mut_ptr.cast()`

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:11:9
   |
LL |         $ptr as *const i32
   |         ^^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `$ptr.cast::<i32>()`
...
LL |     let _ = cast_it!(ptr);
   |             ------------- in this macro invocation
   |
   = note: this error originates in the macro `cast_it` (in Nightly builds, run with -Z macro-backtrace for more info)

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:63:13
   |
LL |     let _ = ptr as *const i32;
   |             ^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `ptr.cast::<i32>()`

error: `as` casting between raw pointers without changing its mutability
  --> $DIR/ptr_as_ptr.rs:64:13
   |
LL |     let _ = mut_ptr as *mut i32;
   |             ^^^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `mut_ptr.cast::<i32>()`

error: aborting due to 8 previous errors

